### PR TITLE
Move row expression above last projectionPushDown

### PR DESCRIPTION
**Pull request by hellium01**
```
Release note
-------
no user affecting change.  
```
On top of https://github.com/prestodb/presto/pull/12982

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java [hunk 1]
```diff
@@ -44,13 +44,11 @@ public static void assertPlan(Session session, Metadata metadata, StatsProvider
     {
         MatchResult matches = actual.getRoot().accept(new PlanMatchingVisitor(session, metadata, statsProvider, lookup), pattern);
         if (!matches.isMatch()) {
-            String formattedPlan = textLogicalPlan(planSanitizer.apply(actual.getRoot()), actual.getTypes(), metadata.getFunctionManager(), StatsAndCosts.empty(), session, 0);
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java]: I believe the unresolved plan is for debugging iterative optimizer. It's worth keeping it still.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java [hunk 2]
```diff
@@ -61,6 +68,12 @@
 
     private static final Pattern<ProjectNode> PATTERN = project()
             .with(source().matching(project().capturedAs(CHILD)));
+    private final FunctionResolution functionResolution;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java]: empty line between constants and variables

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java [hunk 3]
```diff
@@ -84,18 +97,23 @@ public Result apply(ProjectNode parent, Captures captures, Context context)
...
+                .noneMatch(OriginalExpressionUtils::isExpression);
+
         Set<VariableReferenceExpression> inputs = child.getAssignments()
                 .entrySet().stream()
                 .filter(entry -> targets.contains(entry.getKey()))
                 .map(Map.Entry::getValue)
-                .map(OriginalExpressionUtils::castToExpression)
-                .flatMap(entry -> VariablesExtractor.extractAll(entry, context.getVariableAllocator().getTypes()).stream())
+                .flatMap(entry -> isExpression(entry) ? VariablesExtractor.extractAll(castToExpression(entry), context.getVariableAllocator().getTypes()).stream()
+                        : VariablesExtractor.extractAll(entry).stream())
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java]: Use `if/else` if ?/: is too long for a line.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java [hunk 4]
```diff
@@ -141,25 +179,27 @@ private Expression inlineReferences(Expression expression, Assignments assignmen
...
         Set<VariableReferenceExpression> childOutputSet = ImmutableSet.copyOf(child.getOutputVariables());
+        TypeProvider types = context.getVariableAllocator().getTypes();
 
         Map<VariableReferenceExpression, Long> dependencies = parent.getAssignments()
                 .getExpressions()
                 .stream()
-                .map(OriginalExpressionUtils::castToExpression)
-                .flatMap(expression -> VariablesExtractor.extractAll(expression, context.getVariableAllocator().getTypes()).stream())
+                .flatMap(expression -> isExpression(expression) ? VariablesExtractor.extractAll(castToExpression(expression), types).stream()
+                        : VariablesExtractor.extractAll(expression).stream())
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java]: same

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java [hunk 5]
```diff
@@ -141,25 +179,27 @@ private Expression inlineReferences(Expression expression, Assignments assignmen
...
                 .filter(childOutputSet::contains)
                 .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
 
         // find references to simple constants
         Set<VariableReferenceExpression> constants = dependencies.keySet().stream()
-                .filter(input -> castToExpression(child.getAssignments().get(input)) instanceof Literal)
+                .filter(input -> isExpression(child.getAssignments().get(input)) ? castToExpression(child.getAssignments().get(input)) instanceof Literal
+                        : child.getAssignments().get(input) instanceof ConstantExpression)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java]: same

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java [hunk 6]
```diff
@@ -172,6 +212,26 @@ private Expression inlineReferences(Expression expression, Assignments assignmen
         return Sets.union(singletons, constants);
     }
 
+    private Set<VariableReferenceExpression> extractTryArguments(RowExpression expression, TypeProvider types)
+    {
+        if (isExpression(expression)) {
+            return extractTryArguments(castToExpression(expression), types);
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java]: inline `extractTryArguments` for `Expression` into this block.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java [hunk 7]
```diff
@@ -114,6 +117,13 @@
 public class UnaliasSymbolReferences
         implements PlanOptimizer
 {
+    private final FunctionManager functionManager;
+
+    public UnaliasSymbolReferences(FunctionManager functionManager)
+    {
+        this.functionManager = functionManager;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java]: requireNonNull

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 8]
```diff
@@ -105,6 +107,26 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                         identityAssignmentsAsSymbolReferences(restrictedOutputs)));
     }
 
+    /**
+     * @return If the node has outputs not in permittedOutputs, returns an identity projection containing only those node outputs also in permittedOutputs in RowExpression.
+     */
+    public static Optional<PlanNode> addRowExpressionIdentityProjection(PlanNodeIdAllocator idAllocator, PlanNode node, Set<VariableReferenceExpression> permittedOutputs)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Why don't we call it `restrictOutputs`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionVariableInliner.java [hunk 9]
```diff
@@ -31,19 +31,19 @@
         extends RowExpressionRewriter<Void>
 {
     private final Set<String> excludedNames = new HashSet<>();
-    private final Function<VariableReferenceExpression, RowExpression> mapping;
+    private final Function<VariableReferenceExpression, ? extends RowExpression> mapping;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionVariableInliner.java]: Why these changes?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java (hunk 1 above)
Comment (hellium01) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java]: However it will contain `GroupReference` nodes which are not suppose to be in the plan tree. `TranslateExpression` will not be able to run on such plan trees and thus cannot translate the `OriginalExpression`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 8 above)
Comment (hellium01) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: It will conflict with existing one since the signature is same.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionVariableInliner.java (hunk 9 above)
Comment (hellium01) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionVariableInliner.java]: We can have both `Map<VariableReferenceExpression, VariableReferenceExpression>` and `Map<VariableReferenceExpression, RowExpression>`.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java (hunk 1 above)
Comment (highker) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java]: `GroupReference` is for debugging purpose to check if iterative optimizer works. If someone changed `memo`, we need this to debug if the change is correct. I'm ok leaving it out for now; but just be aware of the use case.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java (hunk 6 above)
Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java]: Comment https://github.com/prestodb/presto/pull/13047/commits/2a5f744ca40b7b330ac5056b5b317eab35683a2c#r312617399 was not addressed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 8 above)
Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Put them with similar names to indicate the types of output. The two functions are doing the same thing; they should have similar names.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java (hunk 1 above)
Comment (hellium01) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java]: I will add a TODO to add it back once we have removed all `OriginalExpression`. Also, I think `memo` changes should be captured by `TestMemo` not integration tests.

Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java]: Do we have an issue tracking all the TODOs? I start to worry we might forget some of the clean ups needed as the TODO list is getting long.

File: presto-main/src/main/java/com/facebook/presto/sql/relational/FunctionResolution.java [hunk 10]
```diff
@@ -230,12 +230,12 @@ public boolean isSubscriptFunction(FunctionHandle functionHandle)
 
     public FunctionHandle tryFunction(Type returnType)
     {
-        return functionManager.lookupFunction("TRY", fromTypes(returnType));
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/FunctionResolution.java]: I don't get this change. If `TRY` is registered as `$internal$try`, how could this work before? Otherwise how could this work after?

Comment (hellium01) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/FunctionResolution.java]: We never called this before. Try will always be desugared into a FunctionCall in `DesugarTryExpression`. So we will never actually call `visitTry` in `SqlToRowExpressionTranslator`.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestEliminateSorts.java [hunk 11]
```diff
@@ -87,16 +87,37 @@ public void testNotEliminateSorts()
 
     public void assertUnitPlan(@Language("SQL") String sql, PlanMatchPattern pattern)
     {
-        List<PlanOptimizer> optimizers = ImmutableList.of(
-                new UnaliasSymbolReferences(),
+        List<PlanOptimizer> optimizeBeforeTranslation = ImmutableList.of(
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestEliminateSorts.java]: optimize*r*?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java [hunk 12]
```diff
@@ -172,12 +198,43 @@ private Expression inlineReferences(Expression expression, Assignments assignmen
...
+                if (functionResolution.isTryFunction(call.getFunctionHandle())) {
+                    context.addAll(VariablesExtractor.extractAll(call));
+                }
+                return super.visitCall(call, context);
+            }
+        }, builder);
+        return builder.build();
+    }
+
+    private static List<VariableReferenceExpression> extractDependencies(RowExpression expression, TypeProvider types)
```
[the remark this answers is not in the document]
Comment (hellium01) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineProjections.java]: `extractInput`?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestEliminateSorts.java (hunk 11 above)
Comment (hellium01) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestEliminateSorts.java]: Here, we wanted to test two scenarios:
1) `PruneUnreferencedOutputs`/ `UnaliasSymbolReferences` happens after translateExpressions.
2)  `PruneUnreferencedOutputs`/ `UnaliasSymbolReferences` happens before translateExpressions (it should still work).

Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestEliminateSorts.java]: I mean they should be named `optimizersBeforeTranslation` and `optimizersAfterTranslation`

**Comment by hellium01** [comment]
Thanks for reviewing! Updated to the last comments.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java (hunk 1 above)
Comment (hellium01) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanAssert.java]: Created an issue: https://github.com/prestodb/presto/issues/13231

**Comment by highker** [comment]
Is this good to merge? Assign to me if it is.

**Comment by rongrong** [comment]
Consider that these refactoring has triggered some bugs in previous releases, do we want to run some independent verifier tests before merging?

**Comment by hellium01** [comment]
Yes, verifier tests passed with 4000 queries.
